core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -1,187 +1,337 @@
-        pubkey::Pubkey,
-        timing::AtomicInterval,
-    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey, timing::AtomicInterval},
     std::{
-        collections::{BTreeMap, HashMap},
-        sync::{Arc, Mutex, RwLock},
+        collections::{HashMap, VecDeque},
+        hash::RandomState,
+        ops::Range,
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc, Mutex, RwLock,
+        },
     },
 };
 
 // Limit the size of cluster-slots map in case
 // of receiving bogus epoch slots values.
-const CLUSTER_SLOTS_TRIM_SIZE: usize = 524_288; // 512K
+// This also constraints the size of the datastructure
+// if we are really really far behind.
+const CLUSTER_SLOTS_TRIM_SIZE: usize = 50000;
+
+pub type Stake = u64;
+
+//This is intended to be switched to solana_pubkey::PubkeyHasherBuilder
+type PubkeyHasherBuilder = RandomState;
+pub(crate) type ValidatorStakesMap = HashMap<Pubkey, Stake, PubkeyHasherBuilder>;
 
-pub(crate) type SlotPubkeys = HashMap</*node:*/ Pubkey, /*stake:*/ u64>;
+/// Pubkey-stake map for nodes that have confirmed this slot.
+/// If stake is zero the node has not confirmed the slot.
+pub(crate) type SlotPubkeys =
+    HashMap<Pubkey, AtomicU64 /* stake supporting */, PubkeyHasherBuilder>;
+
+/// Amount of stake that needs to lock the slot for us to stop updating it.
+/// This must be above `DUPLICATE_THRESHOLD` but also high enough to not starve
+/// repair (as it will prefer nodes that have confirmed a slot)
+const FREEZE_THRESHOLD: f64 = 0.9;
+// whatever freeze threshold we should be above DUPLICATE_THRESHOLD in order to not break consensus.
+static_assertions::const_assert!(FREEZE_THRESHOLD > DUPLICATE_THRESHOLD * 1.1);
 
 #[derive(Default)]
 pub struct ClusterSlots {
-    cluster_slots: RwLock<BTreeMap<Slot, Arc<RwLock<SlotPubkeys>>>>,
-    validator_stakes: RwLock<Arc<NodeIdToVoteAccounts>>,
-    epoch: RwLock<Option<u64>>,
-    cursor: Mutex<Cursor>,
-    last_report: AtomicInterval,
+    // ring buffer storing, per slot, which stakes were committed to a certain slot.
+    cluster_slots: RwLock<VecDeque<RowContent>>,
+    // a cache of validator stakes for reuse internally, updated at epoch boundary.
+    validator_stakes: RwLock<Arc<ValidatorStakesMap>>,
+    total_stake: AtomicU64, // total amount of stake across all validators in validator_stakes.
+    current_slot: AtomicU64, // current slot at the front of ringbuffer.
+    epoch: RwLock<Option<u64>>, //current epoch.
+    cursor: Mutex<Cursor>,  // cursor to read CRDS.
+    metrics_last_report: AtomicInterval, // last time statistics were reported.
+    metric_allocations: AtomicU64, // total amount of memory allocations made.
+    metric_write_locks: AtomicU64, // total amount of write locks taken outside of initialization.
+}
+
+#[derive(Debug)]
+struct RowContent {
+    slot: Slot,               // slot for which this row stores information
+    total_support: AtomicU64, // total committed stake for this slot
```

Comment:
Internally VecDeque stores RowContent which includes the total_support for the particular slot to support the lockout mechanism